File: tests/kotlin/test_7_8_1_type_conversion_steps.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.8.1: Type Conversion Traversal Steps.
 *
 * Covers asBool() and asNumber() steps added in TinkerPop 3.8.0.
 * Mirrors tests/python/test_7_8_1_type_conversion_steps.py.
 */
class Test_7_8_1_TypeConversionSteps : StringSpec({

    // ── asBool ────────────────────────────────────────────────────────────────

    "asBool passthrough for Boolean true" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", true)
            val result = g.traversal().V().values<Boolean>("v").asBool().toList()
            result shouldBe listOf(true)
        } finally { g.close() }
    }

    "asBool passthrough for Boolean false" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", false)
            val result = g.traversal().V().values<Boolean>("v").asBool().toList()
            result shouldBe listOf(false)
        } finally { g.close() }
    }

    "asBool from String 'true'" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "true")
            val result = g.traversal().V().values<String>("v").asBool().toList()
            result shouldBe listOf(true)
        } finally { g.close() }
    }

    "asBool from String 'TRUE' is case-insensitive" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "TRUE")
            val result = g.traversal().V().values<String>("v").asBool().toList()
            result shouldBe listOf(true)
        } finally { g.close() }
    }

    "asBool from String 'false'" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "false")
            val result = g.traversal().V().values<String>("v").asBool().toList()
            result shouldBe listOf(false)
        } finally { g.close() }
    }

    "asBool from String '1' is true" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "1")
            val result = g.traversal().V().values<String>("v").asBool().toList()
            result shouldBe listOf(true)
        } finally { g.close() }
    }

    "asBool from String '0' is false" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "0")
            val result = g.traversal().V().values<String>("v").asBool().toList()
            result shouldBe listOf(false)
        } finally { g.close() }
    }

    "asBool from non-zero Int is true" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", 42)
            val result = g.traversal().V().values<Int>("v").asBool().toList()
            result shouldBe listOf(true)
        } finally { g.close() }
    }

    "asBool from zero Int is false" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", 0)
            val result = g.traversal().V().values<Int>("v").asBool().toList()
            result shouldBe listOf(false)
        } finally { g.close() }
    }

    "asBool from unsupported type raises" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", listOf(1, 2, 3))
            val caught = runCatching {
                g.traversal().V().values<Any>("v").asBool().toList()
            }.exceptionOrNull()
            (caught != null) shouldBe true
        } finally { g.close() }
    }

    // ── asNumber ──────────────────────────────────────────────────────────────

    "asNumber passthrough for Int" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", 42)
            val result = g.traversal().V().values<Int>("v").asNumber().toList()
            result shouldBe listOf(42)
        } finally { g.close() }
    }

    "asNumber passthrough for Double" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", 3.14)
            val result = g.traversal().V().values<Double>("v").asNumber().toList()
            result shouldBe listOf(3.14)
        } finally { g.close() }
    }

    "asNumber from String integer returns Long" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "100")
            val result = g.traversal().V().values<String>("v").asNumber().toList()
            result.first() shouldBe 100L
        } finally { g.close() }
    }

    "asNumber from String float returns Double" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "2.718")
            val result = g.traversal().V().values<String>("v").asNumber().toList()
            (result.first() as Double) shouldBe 2.718
        } finally { g.close() }
    }

    "asNumber from Boolean true is 1L" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", true)
            val result = g.traversal().V().values<Boolean>("v").asNumber().toList()
            result shouldBe listOf(1L)
        } finally { g.close() }
    }

    "asNumber from Boolean false is 0L" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", false)
            val result = g.traversal().V().values<Boolean>("v").asNumber().toList()
            result shouldBe listOf(0L)
        } finally { g.close() }
    }

    "asNumber from unparseable String raises" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", "not-a-number")
            val caught = runCatching {
                g.traversal().V().values<String>("v").asNumber().toList()
            }.exceptionOrNull()
            (caught != null) shouldBe true
        } finally { g.close() }
    }

    "asNumber from unsupported type raises" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "v", listOf(1))
            val caught = runCatching {
                g.traversal().V().values<Any>("v").asNumber().toList()
            }.exceptionOrNull()
            (caught != null) shouldBe true
        } finally { g.close() }
    }

    // ── pipeline composition ──────────────────────────────────────────────────

    "asNumber then comparison pipeline" {
        val g = TinkerCat.open()
        try {
            g.addVertex(TinkerVertex.T.label, "e", "score", "95")
            val result = g.traversal().V().values<String>("score").asNumber().toList()
            result.size shouldBe 1
            result.first() shouldBe 95L
        } finally { g.close() }
    }
})

File: tests/python/test_7_8_1_type_conversion_steps.py
"""
Tests for Task 7.8.1: Type Conversion Traversal Steps.

Covers as_bool() and as_number() steps added in TinkerPop 3.8.0.
Mirrors tests/kotlin/test_7_8_1_type_conversion_steps.kt.
"""

import pytest

from tinkercat import TinkerCat


# ── asBool ────────────────────────────────────────────────────────────────────

def test_as_bool_passthrough_true():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=True)
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [True]
    finally:
        g.close()


def test_as_bool_passthrough_false():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=False)
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [False]
    finally:
        g.close()


def test_as_bool_from_string_true():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="true")
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [True]
    finally:
        g.close()


def test_as_bool_from_string_true_case_insensitive():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="TRUE")
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [True]
    finally:
        g.close()


def test_as_bool_from_string_false():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="false")
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [False]
    finally:
        g.close()


def test_as_bool_from_string_one_is_true():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="1")
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [True]
    finally:
        g.close()


def test_as_bool_from_string_zero_is_false():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="0")
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [False]
    finally:
        g.close()


def test_as_bool_from_nonzero_int_is_true():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=42)
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [True]
    finally:
        g.close()


def test_as_bool_from_zero_int_is_false():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=0)
        result = g.traversal().V().values("v").as_bool().to_list()
        assert result == [False]
    finally:
        g.close()


def test_as_bool_from_unsupported_type_raises():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=[1, 2, 3])
        with pytest.raises((TypeError, ValueError, Exception)):
            g.traversal().V().values("v").as_bool().to_list()
    finally:
        g.close()


# ── asNumber ──────────────────────────────────────────────────────────────────

def test_as_number_passthrough_int():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=42)
        result = g.traversal().V().values("v").as_number().to_list()
        assert result == [42]
    finally:
        g.close()


def test_as_number_passthrough_float():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=3.14)
        result = g.traversal().V().values("v").as_number().to_list()
        assert result == [3.14]
    finally:
        g.close()


def test_as_number_from_string_integer():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="100")
        result = g.traversal().V().values("v").as_number().to_list()
        assert result == [100]
    finally:
        g.close()


def test_as_number_from_string_float():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="2.718")
        result = g.traversal().V().values("v").as_number().to_list()
        assert abs(result[0] - 2.718) < 1e-9
    finally:
        g.close()


def test_as_number_from_bool_true_is_1():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=True)
        result = g.traversal().V().values("v").as_number().to_list()
        assert result == [1]
    finally:
        g.close()


def test_as_number_from_bool_false_is_0():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=False)
        result = g.traversal().V().values("v").as_number().to_list()
        assert result == [0]
    finally:
        g.close()


def test_as_number_from_unparseable_string_raises():
    g = TinkerCat()
    try:
        g.add_vertex("e", v="not-a-number")
        with pytest.raises((ValueError, Exception)):
            g.traversal().V().values("v").as_number().to_list()
    finally:
        g.close()


def test_as_number_from_unsupported_type_raises():
    g = TinkerCat()
    try:
        g.add_vertex("e", v=[1, 2])
        with pytest.raises((TypeError, ValueError, Exception)):
            g.traversal().V().values("v").as_number().to_list()
    finally:
        g.close()


# ── pipeline composition ──────────────────────────────────────────────────────

def test_as_number_then_comparison():
    g = TinkerCat()
    try:
        g.add_vertex("e", score="95")
        result = g.traversal().V().values("score").as_number().to_list()
        assert len(result) == 1
        assert result[0] > 90
    finally:
        g.close()
